### ISSUE TITLE
Multiple issues with `spack buildcache` command in v0.17

### ISSUE THREAD

**Issue by birnbera** [LLNL/spack#30410]
### Steps to reproduce

Fresh checkouts of Spack `v0.17.2` seem to have multiple issues with build caches. At a minimum, `spack buildcache list` seems to expect the `spack-public` mirror to have a `build_cache` sub-directory/prefix, which when missing generates an unexpected error:

```console
$ git clone -b v0.17.2 https://github.com/spack/spack.git spack
$ . spack/share/spack/setup-env.sh
$ spack -d buildcache list
==> [2022-04-29-17:01:45.861183] Imported buildcache from built-in commands
==> [2022-04-29-17:01:45.865924] Imported buildcache from built-in commands
==> [2022-04-29-17:01:45.867123] Reading config file /local_scratch/spack/etc/spack/defaults/config.yaml
==> [2022-04-29-17:01:45.903219] Reading config file /local_scratch/spack/etc/spack/defaults/mirrors.yaml
==> [2022-04-29-17:01:45.936087] Fetching index from https://mirror.spack.io/build_cache/index.json
==> [2022-04-29-17:01:45.977735] Error: Multiple errors during fetching:
        Error 1: RuntimeError: Unable to read index hash https://mirror.spack.io/build_cache/index.json.hash due to SpackWebError: Download failed: HTTP Error 404: Not Found
        Error 2: RuntimeError: Unable to read index https://mirror.spack.io/build_cache/index.json due to SpackWebError: Download failed: HTTP Error 404: Not Found
Traceback (most recent call last):
  File "/local_scratch/spack/bin/spack", line 100, in <module>
    sys.exit(spack.main.main())
  File "/local_scratch/spack/lib/spack/spack/main.py", line 900, in main
    return _main(argv)
  File "/local_scratch/spack/lib/spack/spack/main.py", line 883, in _main
    return _invoke_command(command, parser, args, unknown)
  File "/local_scratch/spack/lib/spack/spack/main.py", line 553, in _invoke_command
    return_val = command(parser, args)
  File "/local_scratch/spack/lib/spack/spack/cmd/buildcache.py", line 993, in buildcache
    args.func(args)
  File "/local_scratch/spack/lib/spack/spack/cmd/buildcache.py", line 582, in listspecs
    specs = [s for s in specs if s.satisfies(arch)]
UnboundLocalError: local variable 'specs' referenced before assignment
```

This may be addressed, in part, by a combination of #30045 and #30046, however generating an empty list of mirrors by removing the `spack-public` mirror altogether surfaces a slightly different issue:

```console
$ git checkout bugfix/no-error-missing-cache-index # or bugfix/buildcache-error-handling
$ echo 'mirrors:: {}' > ~/.spack/mirrors.yaml
$ spack -d buildcache list
==> [2022-04-29-17:00:22.006375] Imported buildcache from built-in commands
==> [2022-04-29-17:00:22.011212] Imported buildcache from built-in commands
==> [2022-04-29-17:00:22.012411] Reading config file /local_scratch/spack/etc/spack/defaults/config.yaml
==> [2022-04-29-17:00:22.047079] Reading config file /local_scratch/spack/etc/spack/defaults/mirrors.yaml
==> [2022-04-29-17:00:22.048832] Reading config file /compbio/home/andrew.birnberg/.spack/mirrors.yaml
Traceback (most recent call last):
  File "/local_scratch/spack/bin/spack", line 100, in <module>
    sys.exit(spack.main.main())
  File "/local_scratch/spack/lib/spack/spack/main.py", line 900, in main
    return _main(argv)
  File "/local_scratch/spack/lib/spack/spack/main.py", line 883, in _main
    return _invoke_command(command, parser, args, unknown)
  File "/local_scratch/spack/lib/spack/spack/main.py", line 553, in _invoke_command
    return_val = command(parser, args)
  File "/local_scratch/spack/lib/spack/spack/cmd/buildcache.py", line 993, in buildcache
    args.func(args)
  File "/local_scratch/spack/lib/spack/spack/cmd/buildcache.py", line 576, in listspecs
    specs = bindist.update_cache_and_get_specs()
  File "/local_scratch/spack/lib/spack/spack/binary_distribution.py", line 1607, in update_cache_and_get_specs
    binary_index.update()
  File "/local_scratch/spack/lib/spack/spack/binary_distribution.py", line 374, in update
    raise FetchCacheError(fetch_errors)
  File "/local_scratch/spack/lib/spack/spack/binary_distribution.py", line 62, in __init__
    err = errors[0]
IndexError: list index out of range
```

In addition, there seems to be some weirdness around relative paths in `spack buildcache` commands:

```console
$ spack install ninja
$ spack -d buildcache create -d . ninja
==> [2022-04-29-19:01:19.107010] Imported buildcache from built-in commands
==> [2022-04-29-19:01:19.111960] Imported buildcache from built-in commands
==> [2022-04-29-19:01:19.113489] Reading config file /local_scratch/spack/etc/spack/defaults/config.yaml
Traceback (most recent call last):
  File "/local_scratch/spack/bin/spack", line 98, in <module>
    sys.exit(spack.main.main())
  File "/local_scratch/spack/lib/spack/spack/main.py", line 915, in main
    return _main(argv)
  File "/local_scratch/spack/lib/spack/spack/main.py", line 870, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/local_scratch/spack/lib/spack/spack/main.py", line 898, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/local_scratch/spack/lib/spack/spack/main.py", line 550, in _invoke_command
    return_val = command(parser, args)
  File "/local_scratch/spack/lib/spack/spack/cmd/buildcache.py", line 732, in buildcache
    args.func(args)
  File "/local_scratch/spack/lib/spack/spack/cmd/buildcache.py", line 319, in create_fn
    push_url = spack.mirror.push_url_from_directory(args.directory)
  File "/local_scratch/spack/lib/spack/spack/mirror.py", line 674, in push_url_from_directory
    scheme = url_util.parse(output_directory, scheme='<missing>').scheme
  File "/local_scratch/spack/lib/spack/spack/util/url.py", line 79, in parse
    require_url_format(url)
  File "/local_scratch/spack/lib/spack/spack/util/url.py", line 347, in require_url_format
    assert ut is not None
AssertionError
```

However, this works:

```console
$ spack buildcache create -d $(readlink -f .) ninja
```

As an aside, the `spack buildcache update-index` command documentation should be improved to indicate that the `-d` option can take a URL or a mirror name. Currently it requires a URL but it is possible to use a name and is much more convenient to the end user.

### Information on your system

```console
$ spack debug report 
* **Spack:** 0.18.0.dev0 (f98f2d3177b12d5433ff7ee5c4c0f984d76f6ccc)
* **Python:** 3.6.8
* **Platform:** linux-centos7-skylake_avx512
* **Concretizer:** clingo
```

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
This seems to be fixed in Spack v0.18.0:
```console
HEAD is now at 4c6564f10a update changelog for v0.18.0 (#30905)
$ spack buildcache list
==> 0 cached builds.
==> You can query all available architectures with:
  spack buildcache list --allarch
```
while it is reproducible in Spack v0.17.2

**Comment by alalazo** [comment]
I'll be adding the `wontfix` label to this issue, and leave it as a known bug for the v0.17 series. That would allow us to proceed with a v0.17.3 release incorporating the bugfixes that have been backported already.
